### ISSUE TITLE
Second `search()` with `\v$` pattern and `c` flag finds another new line

### ISSUE THREAD

**Issue by backdround** [vim/vim#13755]
### Steps to reproduce

1. Create init.vim:
```vim
function! EchoLine()
  echom(line('.') . " " . col('.'))
endfunction

function! Found()
  call EchoLine()
  call search('\v$', 'nc', v:null, v:null, function('EchoLine'))
endfunction

function! Search()
  call search('\v$', 'nc', v:null, v:null, function('Found'))
endfunction
```
2. Launch vim with the init.vim:
```bash
./src/vim -u ./init.vim ./init.vim
```
3. Call search()
```vim
:call Search()
```
4. See two different positions:
```
1 21
2 36
```

### Expected behaviour

Search of the new line pattern (`\v$`) with the flag "accept current" (`c`) should accept the current new line even in the `skip` function. It's unexpected that the second search with `c` flag finds the next new line symbol

My main target was to find `start` and `end` of the given pattern and then decide should i search further.

### Version of Vim

9.0 (2022 Jun 28, compiled Dec 23 2023 02:25:07) Included patches: 1-2184

### Environment

Linux: Archlinux 6.6.1-arch1-1
Terminal: Kitty
TERM: tmux-256color
Shell: zsh

**Comment by backdround** [comment]
By the way, Is it ok that `col('.')` returns the column that isn't placeable by `cursor()` in the normal mode?

For example, if I search `\v$` on the line:
```
end
```
`col('.')` gives me `4`, which isn't a placeable column. The rightest placeable column is `3`

**Comment by chrisbra** [comment]
I think this is expected (more or less):

`:h search()` says this about the Skip expression:

> If the {skip} expression is given it is evaluated with the cursor positioned on the start of a match.  If it evaluates to non-zero this match is skipped.  This can be used, for example, to skip a match in a comment or a string. {skip} can be a string, which is evaluated as an expression, a function reference or a lambda.

So Vim will try to position the cursor at the start of the match. But since you cannot position the cursor at the position of `$` it's probably be put on the next line instead (I haven't checked in the code about this).

**Comment by backdround** [comment]
`search` with `$` pattern places the cursor on the column one symbol past the last symbol on the current line. As I understand it's literally the end of the line symbol. You can legally place there the cursor in `visual`, `operator-pending` and `insert` mode (not `normal` mode).

Example:
```lua
a = b
--   ^
--   cursor can be placed here in visual mode
```

If you `search` `$` in visual mode from that place (the next column after the last symbol), it doesn't change the placement in `skip` function, but it does change the placement if `search` performs in `skip` function (even with `c` flag). That is what the issue about (though it uses `normal` mode to demonstrate the inconsistent behevior).

Example:
```lua
-- Initial situation, visual mode on:
a = b
--   ^
a = c

-- Initial situation after two sequential search("\v$", "c").
-- Nothing changes:
a = b
--   ^
a = c

-- Initial situation after search("\v$", "c", v:null, v:null, function('The_same_search')).
-- The cursor moves to the second line:
a = b
a = c
--   ^
```
As I understand these two situations doesn't differ one from another, but works differently

**Comment by chrisbra** [comment]
That's what I have been saying, you cannot place the cursor there. Well, internally the search() function places the cursor there, but it detects that the cursor is on the NULL of the line and it invalidates that match and goes on to the next match.

**Comment by backdround** [comment]
It seems to me that if I perform any `search` with the flag `accept current` twice it should stuck at the second call.

But now it doesn't work that way
Now in visual mode:
```lua
a = b
-- ^
a = c

-- after the first search("\v$", "c") or search("\v$", "ce")
a = b
--   ^
a = c

-- after the second search("\v$", "c") or search("\v$", "ce")
a = b
--
a = c
--   ^
```

What I see is:
- I: place the cursor at the pattern (or the end of the pattern)
- vim: *places at the pattern*
- I: place the cursor at the pattern (or the end of the pattern) and accept it if it's under the cursor (`c`)
- vim: *ignores the current pattern and jumps to the next*

If we can place the cursor on the `NULL` in visual mode (or logically in normal mode), then why we can't accept it?

**Comment by backdround** [comment]
I've been trying to make up a robust algorithm for two days that allows me to get two positions - pattern-start and pattern-end, regardless of a given pattern. With current non-stucking `search` behavior with `^` and `$` patterns, I can't come with something up.

Sorry to ask, but is it possible in the current situation?

**Comment by chrisbra** [comment]
well, pattern start should always work, pattern end should work with the `e` flag for the search() function. Your problem I think is specifically, that Vim does validate the cursor position when searching again in the `skip` expression. Why are you using the `skip` expression for this?
Not sure, what exactly you are trying to achieve, but you may also be using this:
`:%s/pattern/\=MyFunc()/n`

Inside `MyFunc()` you should be able to get the current position (`getpos('.')`) and the end of the match by using `strlen(submatch(0))` I think.

**Comment by backdround** [comment]
I'll reiterate once more:
```lua
-- I manually places the cursor here in visual mode
a = b
--   ^
b = c
--

-- I hit a hotkey with search("\\v$", "c")
a = b
--
b = c
--   ^
```
I don't use `skip` here

Should it jump to the next end of the line and not accept the current (though with `c` flag)?

**Comment by backdround** [comment]
While I tried to workaround the behavior I came across another thing that seems strange to me:

```lua
-- I manually place the cursor here in visual mode
a = b
--^

-- I hit a hotkey with search("\v$", "e")
a = b
--   ^
```
It seems right!

I change the pattern `\v$` => `\vb$`:
```lua
-- I manually places the cursor here in visual mode
a = b
--^

-- I hit a hotkey with search("\vb$", "e")
a = b
--  ^
```
It seems strange for me because all I've done is added a "b" and the result shrinked:
```lua
-- after search("\v$", "e")
a = b
--   ^

-- after search("\vb$", "e")
a = b
--  ^
```

**Comment by backdround** [comment]
Thank you for your attempt to help. I understand that my obsession with the `end of the line symbol` is annoying. I'm trying to create an iterator over a user given pattern that gives an ability to iterate over matches forward and backward.

Random example:
```go
fileHash := fs.GetFileHa┃sh(c.OutputPath)
// -^  ^--------------^   ^--------------
// previous()  <-  current("hash")  ->

if bytes.Equal(oldOutputFileHash, newOutputFileHash) {
// -------------------------^  ^---------------^  ^---
//            ->            next()      ->     next()
```

Hypothetical simplified plugin usage:
```lua
local function jump_to_end_of_line()
  local iterator = get_iterator("\\v$")

  iterator.end_position() == nil then
    -- The cursor isn't placed on the end of the line
    iterator = iterator.next()
  end

  if iterator.end_position() ~= nil then
    iterator.end_position().set_cursor()
  end
end

-- User usage
vim.keymap.set({ "n", "x", "o" }, "s", jump_to_end_of_line, {})
```

And my main problem is to deal with `$`. That's why I'm so annyoing

<details><summary>current() tradeoff implementation</summary>

```lua
M.current = function(pattern)
  -- Get start_pattern_position
  local start_pattern_position = nil
  vim.fn.search(pattern, "bcWn", nil, nil, function()
    start_pattern_position = position.from_cursor(true)
  end)

  if start_pattern_position == nil then
    return nil
  end

  -- Get end_pattern_position
  local initial_position = position.from_cursor(true)

  start_pattern_position.set_cursor()
  local end_pattern_position = nil

  -- Searches the end if the cursor is palced at "$"
  vim.fn.search(pattern, "bceWn", nil, nil, function()
    end_pattern_position = position.from_cursor(true)
  end)

  if start_pattern_position ~= end_pattern_position then
    -- Searches the end in all other cases barring "$"
    vim.fn.search(pattern, "ceWn", nil, nil, function()
      end_pattern_position = position.from_cursor(true)
    end)
  end

  initial_position.set_cursor()

  -- Check that the cursor is in bounds of the current pattern
  if
    start_pattern_position > initial_position
    or end_pattern_position < initial_position
  then
    return nil
  end

  return {
    start_position = start_pattern_position,
    end_position = end_pattern_position,
  }
end
```

</details>

Sorry for `lua`, It's hard for me to write on `vimscript`

**Comment by chrisbra** [comment]
I am not going to try to understand your lua code. I don't know lua and I don't know the neovim api. Please keep it simple for me. 

Are you potentially looking for `:set virtualedit=onemore`? This allows you to place the cursor on the end-of-line. Also not sure what you mean by iterating. Please come up with simple examples.

**Comment by backdround** [comment]
> I am not going to try to understand your lua code. I don't know lua and I don't know the neovim api. Please keep it simple for me.

I tried to give more context, but ok, I won't use lua

> Are you potentially looking for `:set virtualedit=onemore`? This allows you to place the cursor on the end-of-line. Also not sure what you mean by iterating.

No I'm not looking for `:set virtualedit=onemore`. In clean compiled vim with `-u NONE` you already can place the cursor at the end of the line in visual mode with default `:set virtualedit=`.

>  Please come up with simple examples.

I showed the [first](https://github.com/vim/vim/issues/13755#issuecomment-1869224004), [second](https://github.com/vim/vim/issues/13755#issuecomment-1869227510) and [third](https://github.com/vim/vim/issues/13755#issuecomment-1868497172) simple examples.

**Comment by chrisbra** [comment]
that was all about using skip expressions. I find it hard to follow, what exactly you are trying and what is not working. Note: Placing the cursor at the end of the line (after where actually text is, is usually not allowed and I am not going to change that behaviour). 
Would it help, to have `searchpos()` return start and end position of the matc?
